DAQ/src/StrawRecoFromArtdaqFragments_module.cc

```diff
@@ -0,0 +1,297 @@
+
+   // Collection of StrawDigis for the event
+  std::unique_ptr<mu2e::StrawDigiCollection> straw_digis(new mu2e::StrawDigiCollection);
+  std::unique_ptr<mu2e::StrawDigiADCWaveformCollection> straw_digi_adcs(
+                 new mu2e::StrawDigiADCWaveformCollection);
+
+  // IntensityInfoTrackerHits
+  std::unique_ptr<mu2e::IntensityInfoTrackerHits> intInfo(new mu2e::IntensityInfoTrackerHits);
+
+  // FIXME! this is temporary
+
+  std::unique_ptr<mu2e::ProtonBunchTime> pbt(new mu2e::ProtonBunchTime);
+  pbt->pbtime_ = 0;
+  pbt->pbterr_ = 0;
+  event.put(std::move(pbt));
+
+  int np_per_hit  = 1+nADCPackets_;
+  int nsamples    = 3+12*nADCPackets_;
+
+  artdaq::Fragments    fragments;
+  artdaq::FragmentPtrs containerFragments;
+
+  auto fragmentHandles = event.getMany<std::vector<artdaq::Fragment>>();
+
+  for (auto handle : fragmentHandles) {
+    if (!handle.isValid() || handle->empty())     continue;
+
+    if (handle->front().type() == artdaq::Fragment::ContainerFragmentType) {
+      for (const auto& cont : *handle) {
+        artdaq::ContainerFragment contf(cont);
+        // if (contf.fragment_type() != mu2e::FragmentType::DTCEVT) {
+        //   break;
+        // }
+
+        for (size_t ii = 0; ii < contf.block_count(); ++ii) {
+          containerFragments.push_back(contf[ii]);
+          fragments.push_back(*containerFragments.back());
+        }
+      }
+    }
+    else {
+//-----------------------------------------------------------------------------
+// the 'handle' handles a list of artdaq fragments
+// each artdaq fragment corresponds to a single DTC, or a plane
+// loop over them
+//-----------------------------------------------------------------------------
+      int nfrag = handle->size();
+      for (int ifrag=0; ifrag<nfrag; ifrag++) {
+        const auto& frag = handle->at(ifrag);
+        uint8_t* fdata  = (uint8_t*) (frag.dataBegin());
+        TLOG(TLVL_DEBUG) << "-- fragment number:" << ifrag;
+//-----------------------------------------------------------------------------
+// skip non-tracker fragments
+//-----------------------------------------------------------------------------
+        DTCLib::DTC_SubEventHeader* seh = (DTCLib::DTC_SubEventHeader*) fdata;
+        if (seh->source_subsystem != DTCLib::DTC_Subsystem::DTC_Subsystem_Tracker) continue;
+        int dtc_id  = seh->source_dtc_id;
+//-----------------------------------------------------------------------------
+// this is a tracker DTC fragment, loop over the ROCs
+//-----------------------------------------------------------------------------
+        ushort*  buf          = (ushort*) fdata;
+        int      nbytes       = buf[0]; // frag.dataSizeBytes() includes extra 0x20
+        uint8_t* roc_data     = fdata+sizeof(*seh);
+        uint8_t* last_address = fdata+nbytes;     //
+
+        while (roc_data < last_address) {
+          RocDataHeaderPacket_t* rdh = (RocDataHeaderPacket_t*) roc_data;
```

Comment:
Related to my other comment: is there a magic byte, or something, in this header that we can use a flag that there was an error at some point and our we've gone out-of-sequence?

Reply:
Following offline discussion, a consistency check will be implemented at the hit-block level.